[Backport][v1.71.x][Python] Pin Cython to 3.1.1 (#39609)

sreenithi opened the pull request:
Backport of #39609 to v1.71.x.
---
This solves several issues:
1. #39588: Selecting unstable Cython releases when building gRPC Python artifacts.
2. Using inconsistent Cython version in released platform-specific wheels within the same grpcio release.
3. cython/cython#6878 Several published grpcio wheels were built with Cython 3.1.0 and are affected by Cython memory leak issue when using AsyncIO APIs (`grpc.aio.*`).

This change will be backported to 1.73.x, 1.72.x and 1.71.x.

Note: bazel Cython version wasn't upgraded because the latest cython BCR module version is still `3.0.11-1`.

sergiitk commented on tools/dockerfile/grpc_artifact_python_manylinux2014_aarch64.current_version:
```diff
@@ -1 +1 @@
-us-docker.pkg.dev/grpc-testing/testing-images-public/grpc_artifact_python_manylinux2014_aarch64:9e30b45140ef4e1641262913d0df200b6c18b5e8@sha256:263ea79d940c905233624b02751194474408338b0f5ae2fb822966b2f20f47ce
\ No newline at end of file
+us-docker.pkg.dev/grpc-testing/testing-images-public/grpc_artifact_python_manylinux2014_aarch64:8fbdbdf24183fc04c7a261f657b74fe8f72ad238@sha256:04efea9e05cefcdfbb03d46891b67b311a598b42ddf0d3cb8c59ca98802be572
```
Note: a NL added here. I think these .current_version files must be without the nl.  Very annoying. Let's rollback this. You can run `truncate -s -1 tools/dockerfile/grpc_artifact_python_manylinux2014_aarch64.current_version` to fix it.

sergiitk replied:
Looks like this applies to other files too. Here, made you a helper script to fix it in bulk

```sh
find tools/dockerfile/ -name "*.current_version" | xargs -I[] sed -zi '$ s/\n//g' []
```

sergiitk replied:
Not visible in the comment view. Here's what I mean. This PR:
![image](https://github.com/user-attachments/assets/d414bcab-094b-46fe-a1ed-1717e32a3a23)

Original PR:
![image](https://github.com/user-attachments/assets/f46c5dde-e096-447b-9bd5-2adc45536333)

I the tooling works without that last nl, we should add it in all `.current_version` files - but not in a backport to a branch.

sreenithi replied:
oh, thanks for catching this. Looks like the new line crept in while resolving merge conflicts again. The shell command was very helpful. Thank you!

sergiitk reviewed (approved):
Actually, I'll approve it - just run that command to remove the newlines and make sure to squash commits when merging

sreenithi commented:
Addressed all comments, the newline is fixed now.
![image](https://github.com/user-attachments/assets/72205b8c-ee55-49c9-a7c3-a41dc9ea42d1)

But looks like I don't have merge permissions for this one now
